### PR TITLE
Fix block LDLT factorization with small exact entries

### PR DESCRIPTION
**Pull request by orlitzky**
(Follow-up to https://github.com/sagemath/sage/pull/42549)

This makes the `block_ldlt()` algorithm more robust with regard to small exact entries by performing some exact tests when underflow may have happened in an important branch.

Work items:

* Rebase after https://github.com/sagemath/sage/pull/42549 is merged.

### CODE REVIEW

**Review by cxzhong — COMMENTED** [review]
I found two blocking regressions while investigating the CI failures. Details and suggested fixes are inline.

File: src/sage/matrix/matrix2.pyx [hunk 1]
```diff
@@ -15212,18 +15235,30 @@ cdef class Matrix(Matrix1):
...
+            if a_kk_abs*omega_r >= alpha*(omega_1**2):
                 # Step (2) in Higham or Step (4) in B&K.
-                d.append(one_by_one_space(A_kk))
-                _block_ldlt_pivot1x1(A, k)
-                k += 1
-                continue
+                #
+                # Note: we can miss this case if the approximate
+                # a_kk_abs is zero but the real entry is non-zero.
+                if not a_kk_abs and A_kk:
```
Comment (cxzhong) [inline src/sage/matrix/matrix2.pyx]: **This guard drops the ordinary Step (4) 1x1 pivot.** Before this change, once the outer inequality held, the code always pivoted on `A_kk`. Now it does so only when the double approximation is zero, so every ordinary case with `a_kk_abs != 0` falls through to the `A_rr`/2x2 cases.

A small exact reproducer is:

```python
A = matrix(QQ, [[1/2, 0, 1],
                [0,   0, 4],
                [1,   4, 2]])
A.det()          # -8
A.block_ldlt()   # ValueError: matrix equation has no solutions
```

The fallthrough chooses the singular block `[[1/2, 1], [1, 2]]`, although a valid factorization starts with the `1/2` 1x1 pivot.

**Suggested fix:** preserve the original Step (4) branch and use the exact entry only as the zero guard, for example:

```python
if a_kk_abs*omega_r >= alpha*(omega_1**2) and A_kk:
    d.append(one_by_one_space(A_kk))
    _block_ldlt_pivot1x1(A, k)
    k += 1
    continue
```

The earlier Step (2) 1x1 branch should likewise guard `A_kk` if it must handle the spurious nonzero approximation of exact QQbar zero. Please also add the matrix above as a regression test.

File: src/sage/rings/qqbar.py [hunk 2]
```diff
@@ -3967,6 +3967,10 @@ def __abs__(self):
             sage: v.abs().minpoly()
             x - 1
         """
+        if self.is_zero():
```
Comment (cxzhong) [inline src/sage/rings/qqbar.py]: **An unconditional exact zero test makes `abs()` pathologically slow.** `is_zero()` can require a full exact computation for zero or near-zero algebraic numbers (as the `__bool__` documentation below notes). This is the cause of the current CI failures: all seven failing Conda jobs time out in `graded_ring_element.py` at

```python
abs(aut_fact - f_rho.parent().aut_factor(A, z)) < 1e-12
```

The same test times out locally at this call. For the identical operand, the previous descriptor path

```python
AlgebraicReal(u._descr.abs(u)) < 1e-12
```

finishes normally, while the new `abs(u)` does not finish within 10 seconds.

**Suggested fix:** do not call `self.is_zero()` for every algebraic absolute value. Keep the exact-zero check local to the `block_ldlt` pivot decisions where a false nonzero would cause division by zero (guard the relevant 1x1 pivots with `A_kk`), or fix the specific zero descriptor/approximation without forcing general exactification. The `graded_ring_element.py` expression above should be added as a performance regression test (or at least included in the targeted doctest run).

File: src/sage/matrix/matrix2.pyx (hunk 1 above)
Comment (orlitzky) [reply src/sage/matrix/matrix2.pyx]: Ah, thanks, this is just a "typo." I wrote the condition wrong, it should be `if a_kk_abs or A_kk`. The goal is to avoid testing `A_kk` when we already know that it is non-zero (because `a_kk_abs` is non-zero). In most cases `a_kk_abs` will be non-zero, so we avoid the expensive test.

File: src/sage/rings/qqbar.py (hunk 2 above)
Comment (orlitzky) [reply src/sage/rings/qqbar.py]: I'll have to think about this.

There are undoubtedly many other places in the sage library where it is important to know whether or not an algebraic number is zero. It would not make sense to update all of those locations with extra tests, especially since the pending calcium PR fixes the underlying issue.

If I can't think of another way around it, I would rather leave this bug alone, and wait for calcium to fix it. `float(abs(zero))` being non-zero is itself a bug that must be fixed one way or another, and then the `block_ldlt()` problem goes away.

Comment (orlitzky) [reply src/sage/rings/qqbar.py]: I have removed the controversial qqbar.py changes, leaving only the tweaks to handle small exact (rational, in the examples) entries.

**Comment by cxzhong** [comment]
I found a blocking performance regression in the [new exact fallback](https://github.com/sagemath/sage/blob/ca38b06eb64ef6dde1dba5d7fdb40288da4a6415/src/sage/matrix/matrix2.pyx#L15182-L15199). When every approximate subdiagonal magnitude is zero, `if a_ik_abs` forces an exact zero test. Proving that a lazy `AA` or `QQbar` expression is zero can be extraordinarily expensive.

Here is a small reproducer:

```python
a = sum(AA(p).sqrt() for p in primes_first_n(7))
z = ((a*a).sqrt() - a) / 2**2000
assert float(abs(z)) == 0.0
A = matrix(AA, [[0, z], [z, 0]])
A.block_ldlt()
```

The value `z` is exactly zero because `a` is positive and `sqrt(a^2) = a`. On the same machine and environment, the current `develop` implementation returns from `block_ldlt()` in about 0.0013 seconds. At `ca38b06`, it did not return within 20 seconds. Using six square roots instead of seven completes, but takes about 6.7 seconds.

This is also relevant to #42710. That PR correctly rounds lazy exact zeros to `0.0`, which makes them enter this exact fallback more often. In a combined probe using the lazy algebraic zero from `graded_ring_element.py`, the baseline `block_ldlt()` returned in about 0.035 seconds, while #42665 plus #42710 did not return within 20 seconds.

The correctness goal of distinguishing an underflowed nonzero entry from zero is clear, but this changes a millisecond-scale factorization of an exact zero matrix into an effectively unbounded exactification. I think this needs to be resolved before merge: either preserve the fast approximate-zero path for rings with expensive zero tests and narrow the scope accordingly, or first provide a cheap tri-state zero test/backend. Please also add a performance regression test covering a hard algebraic zero.

**Comment by orlitzky** [comment]
There is no way to fix this in general. If we don't do the exact-zero test, we can get the wrong answer in `AA`, `QQbar`, and elsewhere. For example, on _develop_: 

```
sage: t = AA(1) / 2^2000
sage: A = matrix(AA, [[0, t], [t, 0]])
sage: P, L, D = A.block_ldlt()
sage: P.T * A * P == L * D * L.T
False
```

On the other hand, if we do make the exact zero comparison, then in some cases it will be slow. If you change the ring in your example to `SR`, then the fast float approximation in `AA` no longer incidentally leads to a correct factorization (`A.change_ring(SR).block_ldlt()` has performance issues even on _develop_.)

I'll need to rebase this on your full-row-scan PR anyway, but I don't think we'll be able to avoid it being slow sometimes in rings where `is_zero()` can be slow.

**Comment by orlitzky** [comment]
rebased
